Article : introduction au JSON dans PostgreSQL (#1238)
geotribu/website: content/articles/2025/2025-01-21_travailler-avec-JSON-et-PostgreSQL.md

```

On expliquera ce `->` par la suite.

Pour les aventuriers et aventurières, il existe une extension de PostgreSQL `btree_gin` permettant de faire des index multi-champs mixant des champs classiques et `json`. Elle est disponible nativement à l'installation de PostgreSQL et ne vous demandera pas de devenir développeur⸱se C/C++ pour l'installer (coucou les FDW parquet ! Ça va par chez vous ?).

## Création des tables

Je vais éviter de vous spammer du `DDL` ([Data Definition Language](https://fr.wikipedia.org/wiki/Langage_de_d%C3%A9finition_de_donn%C3%A9es), mais vous pourrez retrouver le schéma complet de la base [ici](https://github.com/thomas-szczurek/base_donnees_insee/tree/main/sql/creation_tables).

En voici cependant un schéma succinct pour aider à la compréhension du reste de l'article :

![Diagramme du modèle de la base de données](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2024/postgresql_json/modele.png){: .img-center loading=lazy }

*Ne vous préoccupez pas des tables empilées tout à droite, ce sont des partitions de la table `donnees_communes`. Le sujet ne sera pas évoqué ici et elles ne sont nécessaires pour l'exemple.*

> Partant d'un schéma nommé `insee`, on va créer deux tables :+1:

Reviewer: ```suggestion
Partant d'un schéma nommé `insee`, on va créer deux tables :
```

fk_base int2 NOT NULL,
annee int2 NOT NULL,
donnees jsonb NOT NULL,
UNIQUE (fk_base, annee, code_commune)
CONSTRAINT fk_donnees_bases FOREIGN KEY fk_base ON insee.bases(pk_id) ON UPDATE CASCADE ON DELETE CASCADE
);

-- Création des index
-- Un index multichamps est déjà créé automatiquement sur (fk_base, annee, code_commune) grace à la clause UNIQUE de la déclaration de table
CREATE INDEX idx_donnees_com_donnees ON insee.donnees_communes USING gin (donnees);
END;
```

Les plus tatillons des DBA d'entre vous auront remarqué ce `CHECK`. "Mais pourquoi qu'il utilise pas varchar(5) ? Il fait vraiment n'importe quoi !" Tout simplement, car cette forme permet d'utiliser un type de texte aux nombres de caractère **réellement** arbitraire (le type `text`) contrairement à varchar(255), tout en pouvant en contrôler le nombre minimum et maximum avec le prédicat `CHECK` (contrairement à varchar qui ne contrôle que le maximum) comme expliqué sur le [wiki Postgres](https://wiki.postgresql.org/wiki/Don%27t_Do_This#Don.27t_use_varchar.28n.29_by_default).

Et on insère quelques lignes dans notre table insee.bases :

> ```sql

Reviewer: ```suggestion
```sql title="Script SQL d'insertion de données" linenums="1"
```

   pop_25_54_ans_retraites_c,' || POP2554_CS7_C || ',
   pop_25_54_ans_autres_c,' || POP554_CS8_C || ',
   population_55_ans_et_plus_c,' || POP55P_C || ',
   pop_55_ans_et_plus_ans_agriculteurs_c,' || POP55P_CS1_C || ',
   pop_55_ans_et_plus_ans_artisants_commercants_chefs_entreprise_c,' || POP55P_CS2_C || ',
   pop_55_ans_et_plus_ans_cadres_prof_intel_sup_c,' || POP55P_CS3_C || ',
   pop_55_ans_et_plus_ans_professions_intermediaires_c,' || POP55P_CS4_C || ',
   pop_55_ans_et_plus_ans_employes_c,' || POP55P_CS5_C || ',
   pop_55_ans_et_plus_ans_ouvriers_c,' || POP55P_CS6_C || ',
   pop_55_ans_et_plus_ans_retraites_c,' || POP55P_CS7_C || ',
   pop_55_ans_et_plus_ans_autres_c,' || POP55P_CS8_C,
  E'[\t\n\r]','','g') AS data
  FROM insee.rp_population_import
 )

-- Conversion des chaines de texte en json et insertion dans la table
INSERT INTO insee.donnees_communes("code_commune","annee","fk_base","donnees")
SELECT
    CODGEO,
    2021,
    1,
    jsonb_object(string_to_array(data::text,','))
FROM d
ORDER BY "CODGEO";
```

Ouf.

![donnees_communes](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2024/postgresql_json/donnees_communes.png){: .img-center loading=lazy }

> On veillera à bien grouper les inserts dans cet ordre annee/base/code_commune afin de faciliter la lecture des données par PostgreSQL (si votre table est partitionnée, cela sera facilité).

Author: ```suggestion
On veillera à bien grouper les inserts dans cet ordre annee/base/code_commune afin de faciliter la lecture des données par PostgreSQL.
```

 LATERAL (SELECT
     min(annee) AS premiere,
     max(annee) AS derniere
     FROM insee.donnees_communes WHERE fk_base = c.fk_base AND donnees ? c.clef_json) AS a
ORDER BY fk_base, clef_json;
```

![vm_presence_cles_annees](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2024/postgresql_json/vm_presence_cles_annees.png){: .img-center loading=lazy }

La seule difficulté ici est la présence de [LATERAL](https://docs.postgresql.fr/17/queries-table-expressions.html#QUERIES-FROM). Ce « mot » permet d'utiliser un champ de la requête principale dans une sous-requête placée dans une clause FROM. Les éléments de la sous-requête seront ensuite évalués atomiquement avant d'être joints à la table d'origine. Oui, ce n'est pas très facile à expliquer. Ici, le WHERE de la sous-requête va interroger le champ `donnees` de la table "donnees_communes" pour voir s'il y voit la clef_json en train d'être évaluée dans la table "correspondance_clefs_champs" possédant l'alias "c". Si oui, alors on prend la valeur minimum/maximum du champ année et on le joint à cette ligne et uniquement à cette ligne. Puis évaluation de clef_json suivante ... (*évaluée atomiquement*)

> Maintenant, on voudrait proposer un produit un peu plus simple d'utilisation avec tout ça. Excusez-moi d'avance, je copierai une requête de 50 lignes une seconde fois. La seule table utilisée ici que nous n'avons pas créée est une table zonages_administratifs comprenant les codes communes dans un champ "code_admin", et un champ "fk_type" contenant le type de zonage administratif (1 pour les communes).

Reviewer: ```suggestion
Maintenant, on voudrait proposer un produit un peu plus simple d'utilisation avec tout ça. Excusez-moi d'avance, je copierai une requête de 50 lignes une seconde fois.  
La seule table utilisée ici que nous n'avons pas créée est une table `zonages_administratifs` comprenant les codes communes dans un champ `code_admin` et un champ `fk_type` contenant le type de zonage administratif (1 pour les communes).
```

> ```sql

Reviewer: ```suggestion
```sql title="Création d'une vue matérialisée" linenums="1"
```

 codes_com AS (
  SELECT
   code_admin
  FROM insee.zonages_administratifs
  WHERE fk_type = 1
 ),
-- Sélections des clefs et unnest par année
 clefs AS (
  SELECT
   pk_id,
   generate_series(premiere_annee_presence, derniere_annee_presence,1) AS annee
  FROM insee.presence_clefs_annees
 ),
-- cross join des clefs + année unnestées et des codes communaux
 tc AS (
 SELECT
  cc.code_admin AS code_com,
  cl.annee AS annee,
  cl.pk_id AS pk_id
 FROM codes_com AS cc
 CROSS JOIN clefs AS cl
 ),
-- Selection finale avec récupération des données
 final AS (
 SELECT
  tc.code_com,
  tc.annee,
  co.fk_base,
  co.clef_json,
  CASE
   WHEN (d.donnees ->> clef_json) IN ('','null','s','nd') THEN NULL
   ELSE ((d.donnees ->> clef_json)::real)
  END AS valeur
 FROM tc
 JOIN
  insee.donnees_communes AS d ON (tc.code_com = d.code_commune AND tc.annee = d.annee)
 LEFT JOIN
  insee.presence_clefs_annees AS co ON tc.pk_id = co.pk_id
 ORDER BY tc.annee, co.fk_base, co.clef_json, tc.code_com
 )
SELECT * FROM final WHERE valeur IS NOT NULL;
```

![vm_presence_cles_annees](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2024/postgresql_json/vm_donnees_olap.png){: .img-center loading=lazy }

> Attention, la création de cette vue matérialisée ou son refresh peut prendre un certain temps si vous avez stocké beaucoup de données (1 heure chez moi pour les 6 volets de 2015 à 2021).

Reviewer: ```suggestion
Attention, la création de cette vue matérialisée ou son rafraîchissement peut prendre un certain temps si vous avez stocké beaucoup de données (1 heure chez moi pour les 6 volets de 2015 à 2021).
```

Enfin, histoire de vivre avec son temps et non comme un vieil ours des cavernes, on va convertir cette vue matérialisée en fichier [parquet](https://parquet.apache.org/).
Et, pour ça, on va utiliser gdal/ogr qui est décidément incroyable.

> ```sh
> ogr2ogr -of parquet donnees_insee.parquet PG:"dbname='insee' shcema='insee' tables='donnees_communes_olap' user='nom_utilisateur' password='votre_mot_de_passe'"
> ```

Reviewer: ````suggestion
```sh title="Export vue au format parquet"
ogr2ogr -of parquet donnees_insee.parquet PG:"dbname='insee' schema='insee' tables='donnees_communes_olap' user='nom_utilisateur' password='votre_mot_de_passe'"
````
```

> ```sh title="Export vue au format parquet"
> ogr2ogr -of parquet donnees_insee.parquet PG:"dbname='insee' schema='insee' tables='donnees_communes_olap' user='nom_utilisateur' password='votre_mot_de_passe'"
> 

Reviewer: ```suggestion
```sh title="Export vue au format parquet"
ogr2ogr -of parquet donnees_insee.parquet PG:"dbname='insee' schema='insee' tables='donnees_communes_olap' user='nom_utilisateur' password='votre_mot_de_passe'"
```

```

## Création des tables

Je vais éviter de vous spammer du `DDL` ([Data Definition Language](https://fr.wikipedia.org/wiki/Langage_de_d%C3%A9finition_de_donn%C3%A9es), mais vous pourrez retrouver le schéma complet de la base [ici](https://github.com/thomas-szczurek/base_donnees_insee/tree/main/sql/creation_tables).

En voici cependant un schéma succinct pour aider à la compréhension du reste de l'article :

![Diagramme du modèle de la base de données](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2024/postgresql_json/modele.png){: .img-center loading=lazy }

*Ne vous préoccupez pas des tables empilées tout à droite, ce sont des partitions de la table `donnees_communes`. Le sujet ne sera pas évoqué ici et elles ne sont nécessaires pour l'exemple.*

Partant d'un schéma nommé `insee`, on va créer deux tables :+1: 
- la première contiendra la liste des *bases* disponibles, les différents volets du recensement ;
- lae seconde permettant de stocker les données.

> Pour rester concentré sur le JSON, on va s'épargner 95% du modèle sous-jacent. On ne gérera donc pas ici les codes communes, etc. En bonus pour celles et ceux voulant tester ce que PostgreSQL a dans le ventre, je proposerai une variante juste en dessous :

Author: ```suggestion
Pour rester concentré sur le JSON, on va s'épargner 95% du modèle sous-jacent. On ne gérera donc pas ici les codes communes, etc. :
```

> ```sql

Reviewer: ````suggestion
```sql title="Script de création des tables" linenums="1"
````

Et on insère quelques lignes dans notre table insee.bases :

```sql
INSERT INTO insee.bases (nom) VALUES
('rp_population'),
('rp_menages'),
('rp_logement'),
('rp_diplomes'),
('rp_activite'),
('rp_emploi')
```

```sql
 annee int2 NOT NULL,
 fk_base int2 NOT NULL,
 donnees jsonb NOT NULL,
 UNIQUE (annee, code_commune, fk_base),
 CONSTRAINT pk_donnees_communes PRIMARY KEY (pk_id, fk_base),
 CONSTRAINT fk_donnees_bases FOREIGN KEY (fk_base) REFERENCES insee.bases (pk_id) ON UPDATE CASCADE ON DELETE CASCADE
) PARTITION BY LIST (fk_base);
CREATE TABLE insee.donnees_communes_fk_1 PARTITION OF insee.donnees_communes FOR VALUES IN (1);
CREATE TABLE insee.donnees_communes_fk_2 PARTITION OF insee.donnees_communes FOR VALUES IN (2);
CREATE TABLE insee.donnees_communes_fk_3 PARTITION OF insee.donnees_communes FOR VALUES IN (3);
CREATE TABLE insee.donnees_communes_fk_4 PARTITION OF insee.donnees_communes FOR VALUES IN (4);
CREATE TABLE insee.donnees_communes_fk_5 PARTITION OF insee.donnees_communes FOR VALUES IN (5);
CREATE TABLE insee.donnees_communes_fk_6 PARTITION OF insee.donnees_communes FOR VALUES IN (6);
END;

> ```

Reviewer: Pourquoi 2 blocs séparés ? C'est pas une erreur ?

Author: @Guts j'étais en train d'enlever la partie sur les partitions sur les conseils de Michel Galien (comme ça j'ai déjà une nouvelle idée d'articles : randoms trucs éléphantesques -> partitions de tables et fdw)

Reviewer: ah pardon je n'avais pas fait attention. Ok ping moi quand tu as ta nouvelle version stp.

Author: done @Guts ! :)

```sql
SELECT
  (donnees ->> 'melons')::int4 AS nb_melons
FROM insee.donnees_communes
WHERE
  donnees ? 'melons'
  AND code_commune = '99999'
  AND annee = 2024
```

Vous pouvez voir que j'utilise l'opérateur `?`; uniquement valable pour les champs `jsonb` et non ceux en simple `json`. En effet, lorsque l'on requête un champ json/jsonb, un retour vous est fait pour l'ensemble des enregistrements de la table, même ceux ne contenant pas la clé. Comprendre que si votre table contient 100 000 enregistrements, mais que seuls 100 contiennent la clé "melons", ne pas spécifier cette clause WHERE vous renverrait 100 000 lignes, dont 99 900 de `NULL`.`?` est un opérateur json permettant de poser la question "la clé est-elle présente au premier niveau du champ json pour cet enregistrement ?", et on ne récupérerait que nos 100 enregistrements contenant la clé "melons".

Je précise que même si vous êtes encore ici, je suppose que vous savez déjà cela. Cependant, la forme `(quelque_chose)::(type)` est un raccourci de PostgreSQL pour faire un `cast` soit convertir une valeur dans un autre type. Avec `->>` la valeur nous est renvoyée sous forme de texte et nous la convertissons en entier.

> ### Avec du json imbriqué

Reviewer: ```suggestion
### Avec du JSON imbriqué
```

 'cle1', 'valeur1',
 'cle2', jsonb_array(
  'foo', 'bar', 'baz')
  )
 )
```

Cette insertion pourrait tout aussi bien s'écrire avec un cast d'une chaine de texte vers du jsonb. Attention, la syntaxe json doit être ici respectée :

```sql
INSERT INTO test.test (donnees) VALUES
('{"cle1": "valeur1", "cle2": ["foo", "bar", "baz"]}'::jsonb)
```

#### Recupération

Pour récupérer une valeur, on utilise la fonction `jsonb_path_query()` qui possède deux paramètres : `le nom du champ` contenant les données json, et le `json_path` vers la valeur à atteindre. Imaginons que nous voulions récupérer la deuxième valeur de la liste contenue dans "cle2" :

```sql
SELECT
 jsonb_path_query(donnees, '$.cle2[1]')
FROM test.test
```

Le `$` désigne le début du chemin json retourné. Nous faisons suivre ce premier symbole par un point pour passer à l'objet suivant. Puis, par le nom de clé suivant et ainsi de suite jusqu'à la clé recherchée, à laquelle nous collons un `[1]` pour la 2ᵉ valeur de la liste (les valeurs commencent à 0).
Pour plus d'informations sur les `json_path`, vous pouvez consulter la [documentation](https://www.postgresql.org/docs/current/functions-json.html#FUNCTIONS-SQLJSON-PATH).

> ## Attaquons nous au recensement
> 
> Bien, maintenant que nous avons essayé d'expliquer tant bien que mal les concepts avec des exemples simples, car il faut bien commencer quelque part, essayons avec quelque chose d'un peu plus volumineux.

Reviewer: ```suggestion
## Attaquons nous au recensement

![logo INSEE](https://cdn.geotribu.fr/img/logos-icones/entreprises_association/INSEE.svg){: .img-thumbnail-left }

Bien, maintenant que nous avons essayé d'expliquer tant bien que mal les concepts avec des exemples simples, car il faut bien commencer quelque part, essayons avec quelque chose d'un peu plus volumineux.
```

CREATE INDEX idx_donnees_com_donnees ON insee.donnees_communes USING gin (donnees);
END;
```

Les plus tatillons des DBA d'entre vous auront remarqué ce `CHECK`. "Mais pourquoi qu'il utilise pas varchar(5) ? Il fait vraiment n'importe quoi !" Tout simplement, car cette forme permet d'utiliser un type de texte aux nombres de caractère **réellement** arbitraire (le type `text`) contrairement à varchar(255), tout en pouvant en contrôler le nombre minimum et maximum avec le prédicat `CHECK` (contrairement à varchar qui ne contrôle que le maximum) comme expliqué sur le [wiki Postgres](https://wiki.postgresql.org/wiki/Don%27t_Do_This#Don.27t_use_varchar.28n.29_by_default).

Et on insère quelques lignes dans notre table insee.bases :

```sql
INSERT INTO insee.bases (nom) VALUES
('rp_population'),
('rp_menages'),
('rp_logement'),
('rp_diplomes'),
('rp_activite'),
('rp_emploi')
```

```sql
 annee int2 NOT NULL,
 fk_base int2 NOT NULL,
 donnees jsonb NOT NULL,
 UNIQUE (annee, code_commune, fk_base),
 CONSTRAINT pk_donnees_communes PRIMARY KEY (pk_id, fk_base),
 CONSTRAINT fk_donnees_bases FOREIGN KEY (fk_base) REFERENCES insee.bases (pk_id) ON UPDATE CASCADE ON DELETE CASCADE

> ) PARTITION BY LIST (fk_base);

Author: ```suggestion
```

> CREATE TABLE insee.donnees_communes_fk_1 PARTITION OF insee.donnees_communes FOR VALUES IN (1);

Author: ```suggestion
```

> CREATE TABLE insee.donnees_communes_fk_2 PARTITION OF insee.donnees_communes FOR VALUES IN (2);

Author: ```suggestion
```

> CREATE TABLE insee.donnees_communes_fk_3 PARTITION OF insee.donnees_communes FOR VALUES IN (3);

Author: ```suggestion
```

> CREATE TABLE insee.donnees_communes_fk_4 PARTITION OF insee.donnees_communes FOR VALUES IN (4);

Author: ```suggestion
```

> CREATE TABLE insee.donnees_communes_fk_5 PARTITION OF insee.donnees_communes FOR VALUES IN (5);

Author: ```suggestion
```

> CREATE TABLE insee.donnees_communes_fk_6 PARTITION OF insee.donnees_communes FOR VALUES IN (6);

Author: ```suggestion
```

> END;

Author: ```suggestion
```

> ```

Author: ```suggestion
```

---
title: Travailler avec du JSON et PostgreSQL
subtitle: Jason et les éléphants
authors:
    - Thomas SZCZUREK-GAYANT
categories:
    - article
comments: true
date: 2025-01-21
description: "Stocker des données au format json dans PostgreSQL, les consulter... et tout ça avec les données du recensement de l'INSEE pour l'exemple. "
icon: simple/postgresql
image:
license: default
robots: index, follow
tags:
    - PostgreSQL

>     - json

Reviewer: À appliquer partout

```suggestion
    - JSON
```

---

# Travailler avec du JSON dans PostgreSQL

> Dans le cadre d'un projet personnel encadré de type "tu l'as fait pour les iris de Lille avec Sqlite, t'as vu que tu sais faire. Bon bah maintenant tu le fait pour toutes les communes de France dans PG", j'ai voulu stocker une bonne partie des données du recensement de l'Insee dans une base PostgreSQL avec des tables multimillésimes. Problème, même au sein d'un même jeu de données, les champs peuvent changer au cours des années et celà empêche de pouvoir dégager une structure de table fixe, ce qui est assez génant vous en conviendrez. La solution ? Passer par des données semi-structurées, soit stocker ces données en JSON dans le champ d'une table. Cet article se veut un condensé de cette expérience.

Author: j'ai juste un doute sur le empêche/empêche. C'est "les champs pouvant changer" donc un et un seul fait qui empêche non ?

> Dans le cadre d'un projet personnel encadré de type "tu l'as fait pour les iris de Lille avec Sqlite, t'as vu que tu sais faire. Bon bah maintenant tu le fait pour toutes les communes de France dans PG", j'ai voulu stocker une bonne partie des données du recensement de l'Insee dans une base PostgreSQL avec des tables multimillésimes. Problème, même au sein d'un même jeu de données, les champs peuvent changer au cours des années et celà empêche de pouvoir dégager une structure de table fixe, ce qui est assez génant vous en conviendrez. La solution ? Passer par des données semi-structurées, soit stocker ces données en JSON dans le champ d'une table. Cet article se veut un condensé de cette expérience.

Reviewer 2: Mmm, oui, en effet, on peut garder "empêche" au singulier

!!! warning
    Ces travaux ont été réalisés avant la sortie de PostgreSQL 17 qui ajoute d'importantes fonctionnalités pour le JSON avec les [`JSON_TABLE`](https://doc.postgresql.fr/17/functions-json.html#FUNCTIONS-SQLJSON-TABLE), elles ne seront pas évoquées ici.

Puisque nous allons parler de json et de données semi-structurées, je me sens dans l'obligation de commencer cet article par un avertissement.

**Le modèle relationnel c'est bon, mangez en, et les contraintes d'intégrités ont été inventées pour de bonnes raisons.**

Cet article ne se veut surtout pas être une invitation à partir en mode yolo sur la gestion des données "c'est bon ya qu'a tout mettre en json" (comme un vulgaire dev qui mettrait tout dans Mongodb diraient les mauvaises langues).

## Le JSON pour les débutant.es

> Pour celles et ceux qui ne connaisent pas le `json`, il s'agit d'un format textuel de représentation des données fonctionnant en partie sur un système de `clé : valeur` qu'on peut voir comme une sorte d'évolution du `xml`.

Author: la ref directe a ajouté normalement

- Le type `json` qui est là pour des raisons historiques et laisser fonctionner des bases qui auraient utilisé ce type dans le passé. Il stocke les informations sous forme textuelle ce qui peu optimisé pour un ordinateur. Il existe toutefois un intérêt à l'utiliser : il permet de retrouver l'information sur l'ordre des clés. Si il est important pour vous de savoir que `nom` est la clef 1 et `prenom` la clef 2, sans avoir à repasser par le nom de la clef, alors il vous faudra passer par le type `json`.

- le type `jsonb`. Le type moderne. Il stocke les informations sous forme binaire et énormément de fonctions sont disponibles en plus de celles du type `json`.

## Les index

Il est possible d'indexer un champ de type `json` / `jsonb` sur ses clés **de premier niveau** et ça se fait avec des index de type `GIN`.

```sql
CREATE INDEX idx_tb_champjson ON tb USING gin (champ_json);
```

Pour indexer des valeurs plus "profondes", il faudra passer par des indexs fonctionnels (index sur des fonctions)

```sql
CREATE INDEX idx_tb_champjson ON tb USING gin (champ_json -> cle_ou_se_situent_les_valeur_a_indexer);
```

(on va expliquer ce `->` par la suite)

> Pour les aventuriers et aventurières, il existe une extention de PostgreSQL `btree_gin` permettant de faire des index multi-champs mixant des champs classiques et json. Elle est disponible nativement à l'installation de PostgreSQL et ne vous demandera pas de devenir développeur.se C/C++ pour l'installer (coucou les fdw parquet ! Ca va par chez vous ?).

Author: Je suis plutôt adepte du . mais "aventurier et aventurières" ajoute un effet comique je trouve.

## Création des tables

> Je vais éviter de vous spammer du DDL, mais vous pourrez retrouver le schéma complet de la base [ici](https://github.com/thomas-szczurek/base_donnees_insee/tree/main/sql/creation_tables)

Author: non pour éviter de me perdre dans l'explication du modèle relationnel mais je peux faire un screen du modèle sous pgmodeler  pour montrer à voir . En plus effectivement ça serait utile car j'évoque quelques tables plus loin.

- On peut requêter la table parent ou les tables enfants.
- Une table doit être partitionnée à sa création, une table déjà existante ne peut pas être convertie. De nouvelles partitions peuvent toutefois êtres créées à volonté.
- Celà peut être interessant pour limiter la taille des scans séquentiels qui se feront sur des tables plus petites que la table *parent*, où pour se faciliter la gestion de tables volumineuses.
- On peut attacher / détacher une partition (qui devient alors une table classique) avec les "mots" `ATTACH PARTITION` / `DETACH PARTITION` y comprit sur une base en production grace à l'option `CONCURENTLY`
- Une clé primaire de table partitionnée **doit** être composite, c'est à dire que son unicité sera vérifiée par la composition de plusieurs champs, et contenir obligatoirement le champ de partitionnement.

(vous aussi lire le contenu de la [documentation à se sujet](https://doc.postgresql.fr/17/ddl-partitioning.html))

A quoi ressemblerai la création de insee.donnees_communes si on la partitionnait selon les différentes sources de données ?

```sql
BEGIN;
CREATE TABLE insee.donnees_communes (

>  pk_id int4 GENERATED BY DEFAULT AS IDENTITY,

Author: Mon prof particulier de PG était très perfectionniste :)

### Exemple simple

On va créer une chaine de texte qui contiendra le contenu de notre json dont les valeurs serons séparées par des virgules `clé1,valeur1, clé2, valeur2`. Cette chaine sera passée dans une fonction `string_to_array()` la transformant en `array` avec comme séparateur des `,` pour séparer les éléments de la chaine de texte vers des éléments de liste, caractère passé en second paramètre de la fonction. Cet `array` sera ensuite envoyé dans la fonction `jsonb_object()`.

```sql
INSERT INTO insee.donnees_communes (code_commune, annee, fk_base, donnees) VALUES
(
  '99999',
  2024,
  1,
  json_object(
    string_to_array('tomates,42,melons,12',',')
    )
)
```

Cette requête encodera cet objet json dans le champs `données` :

```json
{
  "tomate": 42,
  "melons": 12
}
```

Maintenant comment récupérer notre nombre de melons pour le code commune 99999 en 2024 ? Celà se fait grace à des opérateurs spéciaux :

> - `champ_jsonb -> 'clé'` récupère la valeur d'une clé en concervant son type json

Author: conservant ?

Récupérons le dernier millésime du volet population du recensement communal [au format csv sur le site de l'INSEE](https://www.insee.fr/fr/information/8183122) (vous voulez les bases des principaux indicateurs). Pour l'exemple, on utilisera le fichier "Evolution et structure de la population". Il faut tout d'abord nettoyer les noms de champs car l'INSEE indique le millésime systématiquement dans ses noms de champs, ce qui fait que ces derniers changent tous les ans pour un même indicateur.

Les noms de champs commencent tous par P ou C, ceci indique *exploitation principale* (réponses brutes aux questions du recensement) ou *exploitation complémentaire* (croisement de réponses pour établir un indicateur). Les champs provenant de l'exploitation principale et ceux issus de la complémentaire ne doivent pas êtres croisés entre eux. Cette information est évidemmment a conserver mais par choix personnel je préfère la mettre à la fin plutôt qu'au début du nom.

Vous trouverez [ici](https://github.com/thomas-szczurek/base_donnees_insee/blob/main/sql/import/correction_champs_insee.xlsx) un tableur dont le rôle est de s'occuper de tout ça.

> - `STXT` est la fonction qui permet de découper une chaine de texte avec Excel, avec comme arguments `cellule comprenant la chaine initiale` , `position du début de la découpe` , `position de fin de la découpe`
> - A partir du fichier récupéré, il vous suffit de copier coller les noms de champs INSEE sur la première ligne et de récupérer la seconde, qui contiendra les noms de champs formatés.
> - On  remplace les noms de champs du fichier INSEE original.
> - Et on change le nom de cette table temportaire par "rp_population_import.csv"
> 
> Avant d'insérer les données dans notre table, nous allons passer par une table temporaire afin de rendre les données accessibles dans Postgres. Utiliser `COPY` de Postgresql serait fastidieux car il faudrait indiquer la centaine de champs que contient le volet population du recensement dans la commande. Et je n'ai pas honte de dire que j'ai un baobab dans la main à cette idée. Nous sortons donc ce merveilleux logiciel qu'est QGIS, on active les panneaux Explorateur et Explorateur2, on se créé une conection vers la base avec les droits de création, et d'un mouvement gracile du poignet vous glissez le fichier depuis le panneau Explorateur vers la base Postgres dans l'Explorateur2. Laisser la magie opérer.
> 
> Maintenant préparez vous pour peut-être un des INSERT les plus bizarre de votre vie (en tout ça la été pour moi !). Arf. Je me rend compte que si je veux bien faire il faut aussi que j'explique les CTE (c'est très étrange car encore une fois, si vous êtes encore ici vous savez très probablement déjà ce qu'est une CTE).
> 
> CTE veut dire Common Table Expression. C'est une fonctionnalité qui permet, grace à la clause `WITH` d'isoler une sous requête de la requête principale pour rendre tout un peu plus clair, ou de la nommer pour pouvoir la réutiliser à plusieurs endroits sans devoir la réécrire. On peut aussi s'en servir pour faire des requêtes récursives avec `WITH RECURSIVE` et si le sujet vous interesse je vous encurange à aller lire la [magnifique documentation](https://www.postgresql.org/docs/current/queries-with.html#QUERIES-WITH-RECURSIVE) de Postgres a ce sujet.
> 
> On va utiliser la CTE pour concatenner le nom que l'on veut donner à nos clés avec les valeurs contenues dans notre table temporaire dans une chaine séparée par des `,`, qu'on enverra dans une fonction `string_to_array()` puis dans une fonction `jsonb_object()`. On en profiterra pour au passage nettoyer toute tabulation ou retour chariot qui pourrait subsister avec une expression régulière grace a la fonction `regex_replace()`. (ces caractères se notent `\t`, `\n` et `\r`). Cette dernière fonction prend 3 arguments : la chaine de caractère source, le `pattern` recherché, le texte de remplacement. On y ajoute le *drapeau* optionnel `g` afin de remplacer toutes les occurences trouvées.
> 
> Notez que si votre table temporaire possède un nom différent de "rp_population_import" il vous faudra modifier la clause FROM de la CTE.

Reviewer 2: ```suggestion
- `STXT` est la fonction qui permet de découper une chaine de texte avec Excel, avec comme arguments `cellule comprenant la chaine initiale` , `position du début de la découpe` , `position de fin de la découpe`
- À partir du fichier récupéré, il vous suffit de copier-coller les noms de champs INSEE sur la première ligne et de récupérer la seconde, qui contiendra les noms de champs formatés.
- On  remplace les noms de champ du fichier INSEE original.
- Et on change le nom de cette table temporaire par "rp_population_import.csv"

Avant d'insérer les données dans notre table, nous allons passer par une table temporaire afin de rendre les données accessibles dans Postgres. Utiliser `COPY` de Postgresql serait fastidieux car il faudrait indiquer la centaine de champs que contient le volet population du recensement dans la commande. Et, je n'ai pas honte de dire que j'ai un baobab dans la main à cette idée. Nous sortons donc ce merveilleux logiciel qu'est QGIS. On active les panneaux Explorateur et Explorateur2. On crée une connexion vers la base avec les droits de création, et d'un mouvement gracile du poignet, vous glissez le fichier depuis le panneau Explorateur vers la base Postgres dans l'Explorateur2. Laisser la magie opérer.

Maintenant, préparez-vous pour peut-être un des INSERT les plus bizarres de votre vie (en tout cas, ça l'a été pour moi !). Arf. Je me rends compte que si je veux bien faire, il faut aussi que j'explique les CTE (c'est très étrange, car encore une fois, si vous êtes encore ici, vous savez très probablement déjà ce qu'est une CTE).

CTE veut dire Common Table Expression. C'est une fonctionnalité qui permet grâce à la clause `WITH` d'isoler une sous-requête de la requête principale pour rendre tout un peu plus clair. Elle permet également de la nommer pour pouvoir la réutiliser à plusieurs endroits sans devoir la réécrire. On peut aussi s'en servir pour faire des requêtes récursives avec `WITH RECURSIVE` et si le sujet vous intéresse je vous encourage à aller lire la [magnifique documentation](https://www.postgresql.org/docs/current/queries-with.html#QUERIES-WITH-RECURSIVE) de Postgres a ce sujet.

On va utiliser la CTE pour concaténer le nom que l'on veut donner à nos clés avec les valeurs contenues dans notre table temporaire dans une chaine séparée par des `,`. On l'enverra dans une fonction `string_to_array()` puis dans une fonction `jsonb_object()`. On en profitera pour au passage, nettoyer toute tabulation ou retour chariot qui pourrait subsister avec une expression régulière grâce à la fonction `regex_replace()`. (ces caractères se notent `\t`, `\n` et `\r`). Cette dernière fonction prend 3 arguments : la chaine de caractères source, le `pattern` recherché, le texte de remplacement. On y ajoute le *drapeau* optionnel `g` afin de remplacer toutes les occurrences trouvées.

Notez que si votre table temporaire possède un nom différent de "rp_population_import" il vous faudra modifier la clause FROM de la CTE.
```

 )

-- Conversion des chaines de texte en json et insertion dans la table
INSERT INTO insee.donnees_communes("code_commune","annee","fk_base","donnees")
SELECT
    "CODGEO",
    2021,
    1,
    jsonb_object(string_to_array(data::text,','))
FROM d
ORDER BY "CODGEO";
```

Ouf.

![donnees_communes](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2024/postgresql_json/donnees_communes.png){: .img-center loading=lazy }

On veillera à bien grouper les inserts dans cet ordre annee/base/code_commune afin de faciliter la lecture des données par PostgreSQL (si votre table est partitionnée, celà sera facilité).

Maintenant, imaginez que comme l'auteur de ces lignes, vos gros doigts boudinnés cafouillent et glissent sur les touches lors de la rédaction de cette requête, qu'une faute de frappe s'y glisse, et là c'est le drame. Comment modifier le nom d'une clé déjà encodée dans la table ? Avec cette astuce :

```sql
CREATE TABLE example(id int PRIMARY KEY, champ jsonb);
INSERT INTO example VALUES
    (1, '{"nme": "test"}'),
    (2, '{"nme": "second test"}');

UPDATE example
SET champ = champ - 'nme' || jsonb_build_object('name', champ -> 'nme')
WHERE champ ? 'nme'
returning *;
```

> `-` est un opérateur qui permet de supprimer une clé d'un objet json. Pour notre UPDATE on enlève donc notre faute de frappe de l'ensemble de l'objet et on concatène tout le reste avec la construction d'un nouvel objet où on corrige le nom de la clé, et à laquelle on assigne la valeur de la clé en train d'être supprimée (elle est toujours utilisable au moment de l'UPDATE), sur les champs qui la contiennent à l'origine avec `?`.

Reviewer 2: ```suggestion
`-` est un opérateur qui permet de supprimer une clé d'un objet json. Pour notre UPDATE on enlève donc notre faute de frappe de l'ensemble de l'objet. De plus, on concatène tout le reste avec la construction d'un nouvel objet où l'on corrige le nom de la clé. On assigne également la valeur de la clé en train d'être supprimée, elle est toujours utilisable au moment de l'UPDATE, sur les champs qui la contiennent à l'origine avec `?`.
```

## Et maintenant ? Qu'est-ce qu'on fait de ça ?

> Jusqu'à maintenant, nous n'avons travaillé qu'avec le volet population pour son dernier millésime. Imaginons maintenant que nous répétions l'exercice pour les 6 volets et sur plusieurs millésimes et ce sachant qu'au cours du temps, certains champs peuvent apparaitre / disparaitre (changement dans les niveaux de diplômes observés par exemple). Il serait interessant de récupérer une table indiquant la première et la dernière année de présence de chacune des clés. Mettons que lors de ces travaux nous en avons profité pour mettre à jour une table "correspondance_clefs_champs" listant chacune des clés présente et son nom INSEE d'origine (tout du moins, celui que nous avions normalisé).

Reviewer 2: ```suggestion
Jusqu'à maintenant, nous n'avons travaillé qu'avec le volet population pour son dernier millésime. Imaginons maintenant que nous répétions l'exercice pour les 6 volets et sur plusieurs millésimes, et ce, en sachant qu'au cours du temps, certains champs peuvent apparaitre ou disparaitre ; changement dans les niveaux de diplômes observés par exemple. Il serait intéressant de récupérer une table indiquant la première et la dernière année de présence de chaque clé. Mettons que lors de ces travaux, nous en avons profité pour mettre à jour une table "correspondance_clefs_champs" listant chaque clé présentes et son nom INSEE d'origine (tout du moins, celui que nous avions normalisé).
```

```sql
CREATE MATERIALIZED VIEW insee.presence_clefs_annees AS
SELECT
 c.pk_id AS pk_id,
 c.clef_json AS clef_json,
 c.fk_base AS fk_base,
 p.premiere AS premiere_annee_presence,
 d.derniere AS derniere_annee_presence
FROM insee.correspondance_clefs_champs AS c,
 LATERAL (SELECT min(annee) AS premiere FROM insee.donnees_communes WHERE fk_base = c.fk_base AND donnees ? c.clef_json) AS p,
 LATERAL (SELECT max(annee) AS derniere FROM insee.donnees_communes WHERE fk_base = c.fk_base AND donnees ? c.clef_json) AS d
ORDER BY fk_base, clef_json;
```

![vm_presence_cles_annees](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2024/postgresql_json/vm_presence_cles_annees.png){: .img-center loading=lazy }

> La seule difficulté ici est la présence de [LATERAL](https://docs.postgresql.fr/17/queries-table-expressions.html#QUERIES-FROM). Ce "mot" permet d'utiliser un champ de la requête principale dans une sous requête placée dans une clause FROM. Les éléments de la sous requête seront ensuite évalués atomiquement avant d'être joins à la table d'origine. Oui, ce n'est pas très facile à expliquer. Ici, le WHERE de la sous requête va interroger le champ données de la table "donnees_communes" pour voir si il y voit la clef_json actuellement en train d'être évaluée dans la table "correspondance_clefs_champs" possédant l'alias "c". Si oui, alors on prend la valeur minimum/maximum du champ année, et on le joint à cette ligne et uniquement cette ligne. Puis évaluation de clef_json suivante ... (*évalués atomiquement*)
> 
> Maintenant, on voudrait proposer un produit un peu plus simple d'utilisation avec tout ça. Désolé d'avance, je vais copier une requête de 50 lignes une seconde fois. La seule table utilisée ici que nous n'avons pas créé est une table zonages_administratifs comprenant les codes communes dans un champ "code_admin", et un champ "fk_type" contenant le type de zonage administratif (1 pour les communes).

Reviewer 2: ```suggestion
La seule difficulté ici est la présence de [LATERAL](https://docs.postgresql.fr/17/queries-table-expressions.html#QUERIES-FROM). Ce « mot » permet d'utiliser un champ de la requête principale dans une sous-requête placée dans une clause FROM. Les éléments de la sous-requête seront ensuite évalués atomiquement avant d'être joints à la table d'origine. Oui, ce n'est pas très facile à expliquer. Ici, le WHERE de la sous-requête va interroger le champ donné de la table "donnees_communes" pour voir s'il y voit la clef_json en train d'être évaluée dans la table "correspondance_clefs_champs" possédant l'alias "c". Si oui, alors on prend la valeur minimum/maximum du champ année et on le joint à cette ligne et uniquement à cette ligne. Puis évaluation de clef_json suivante ... (*évaluée atomiquement*)

Maintenant, on voudrait proposer un produit un peu plus simple d'utilisation avec tout ça. Excusez-moi d'avance, je copierai une requête de 50 lignes une seconde fois. La seule table utilisée ici que nous n'avons pas créée est une table zonages_administratifs comprenant les codes communes dans un champ "code_admin", et un champ "fk_type" contenant le type de zonage administratif (1 pour les communes).
```

Reviewer 2: Alors, dans ce cas, oui. Avec la petite syntaxe qui va bien `donnes`

 FROM codes_com AS cc
 CROSS JOIN clefs AS cl
 ),
-- Selection finale avec récupération des données
 final AS (
 SELECT
  tc.code_com,
  tc.annee,
  co.fk_base,
  co.clef_json,
  CASE
   WHEN (d.donnees ->> clef_json) IN ('','null','s','nd') THEN NULL
   ELSE ((d.donnees ->> clef_json)::real)
  END AS valeur
 FROM tc
 JOIN
  insee.donnees_communes AS d ON (tc.code_com = d.code_commune AND tc.annee = d.annee)
 LEFT JOIN
  insee.presence_clefs_annees AS co ON tc.pk_id = co.pk_id
 ORDER BY tc.annee, co.fk_base, co.clef_json, tc.code_com
 )
SELECT * FROM final WHERE valeur IS NOT NULL;
```

![vm_presence_cles_annees](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2024/postgresql_json/vm_donnees_olap.png){: .img-center loading=lazy }

Attention, la création de cette vue matérialisée ou son refresh peut prendre un certain temps si vous avez stocké beaucoup de données (1 heure chez moi pour les 6 volets de 2015 à 2021).

Enfin, histoire de vivre avec son temps et non comme un viel ours des cavernes ou comme un data machin de collectivité territoriale (j'y bosse, j'ai le droit), on va convertir cette vue matérialisée en fichier [parquet](https://parquet.apache.org/).

> Et pour ça on va utiliser gdal/ogr qui est décidemment incroyable.

Reviewer 2: ```suggestion
Et, pour ça, on va utiliser gdal/ogr qui est décidément incroyable.
```

```sh
ogr2ogr -of parquet donnees_insee.parquet PG:"dbname='insee' shcema='insee' tables='donnees_communes_olap' user='nom_utilisateur' password='votre_mot_de_passe'"
```

> Et on peut ainsi mettre le fichier sur un espace cloud, comme [ici](https://donnees-insee.s3.fr-par.scw.cloud/donnees_insee_olap.parquet) ! Vous pouvez ensuite sortir votre plus beau générateur de publications Linkedin qui mettra plein d'emojis choupi et faire le cake sur les rezos (imaginez que 90% du contenu de Linkedin doit être fait avec ces trucs qui sont capable de vous générer des publications expliquant que le shape a des avantages sur le geopackage, le tout sur un ton très assuré).

Reviewer 2: ```suggestion
Et, on peut ainsi mettre le fichier sur un espace cloud, comme [ici](https://donnees-insee.s3.fr-par.scw.cloud/donnees_insee_olap.parquet) ! Vous pouvez ensuite sortir votre plus beau générateur de publications Linkedin qui mettra plein d'émojis choupi et faire le cake sur les rezos (imaginez que 90% du contenu de Linkedin doit être fait avec ces trucs capables de vous générer des publications expliquant que le shape a des avantages sur le GeoPackage, le tout sur un ton très assuré).
```

Reviewer 2: Mais le ShapeFile a des avantages sur le GeoPackage ! 😅

Author: Alors, c'est peut-être l'effet conversation écrite, mais si je ce n'esst pas une blague, je suis preneur des cas pour ma ~~confiture~~ culture !

Reviewer: :popcorn:

Reviewer 2: Non, ce n'est pas une blague :D 
En vrai, le GeoPackage a beaucoup d'avantages sur le ShapeFile. Mais celui-ci en a quand même une importante (dans mon cas) : une meilleure interopérabilité (SIG, DAO, Illustration type Illustrator & cie) aujourd'hui, compatibilité historique et une certaine simplicité

Author: merci ;) (désolé pour le pop corn !)

---
title: Travailler avec du JSON et PostgreSQL
subtitle: Jason et les éléphants
authors:
    - Thomas SZCZUREK-GAYANT
categories:
    - article
comments: true
date: 2025-01-21
description: "Stocker des données au format json dans PostgreSQL, les consulter... et tout ça avec les données du recensement de l'INSEE pour l'exemple. "
icon: simple/postgresql
image:
license: default
robots: index, follow
tags:
    - PostgreSQL
    - JSON
---

# Travailler avec du JSON dans PostgreSQL

:calendar: Date de publication initiale : {{ page.meta.date | date_localized }}

> Dans le cadre d'un projet personnel, j'ai voulu stocker une bonne partie des données du recensement de l'Insee dans une base PostgreSQL avec des tables multimillésimes. Problème, au sein d'un même jeu de données, les champs peuvent changer au cours des années et cela empêche de pouvoir dégager une structure de table fixe, ce qui est assez gênant vous en conviendrez. La solution ? Passer par des données semi-structurées, soit stocker ces données en JSON dans le champ d'une table. Cet article se veut un condensé de cette expérience.

Reviewer: Ajout de marqueurs visuels pour alléger la charge cognitive à la lecture et au parcours de l'article.

```suggestion

![logo JSON](https://cdn.geotribu.fr/img/logos-icones/programmation/json.png){: .img-thumbnail-left }

Dans le cadre d'un projet personnel, j'ai voulu stocker une bonne partie des données du recensement de l'Insee dans une base PostgreSQL avec des tables multimillésimes. Problème, au sein d'un même jeu de données, les champs peuvent changer au cours des années et cela empêche de pouvoir dégager une structure de table fixe, ce qui est assez gênant vous en conviendrez. La solution ? Passer par des données semi-structurées, soit stocker ces données en JSON dans le champ d'une table. Cet article se veut un condensé de cette expérience.
```

> !!! warning

Reviewer: ```suggestion
!!! info "Obsolescence non programmée"
```

> Dans le cadre d'un projet personnel encadré de type "tu l'as fait pour les IRIS de Lille avec Sqlite, t'as vu que tu sais faire. Bon bah maintenant tu le fais pour toutes les communes de France dans PG", j'ai voulu stocker une bonne partie des données du recensement de l'Insee dans une base PostgreSQL avec des tables multimillésimes. Problème, même au sein d'un même jeu de données, les champs peuvent changer au cours des années et cela empêchent de pouvoir dégager une structure de table fixe, ce qui est assez gênant vous en conviendrez. La solution ? Passer par des données semi-structurées, soit stocker ces données en JSON dans le champ d'une table. Cet article se veut un condensé de cette expérience.

Author: ```suggestion
Dans le cadre d'un projet personnel encadré de type "tu l'as fait pour les IRIS de Lille avec Sqlite, t'as vu que tu sais faire. Bon bah maintenant tu le fais pour toutes les communes de France dans PG", j'ai voulu stocker une bonne partie des données du recensement de l'Insee dans une base PostgreSQL avec des tables multimillésimes. Problème, même au sein d'un même jeu de données, les champs peuvent changer au cours des années et cela empêche de pouvoir dégager une structure de table fixe, ce qui est assez gênant vous en conviendrez. La solution ? Passer par des données semi-structurées, soit stocker ces données en JSON dans le champ d'une table. Cet article se veut un condensé de cette expérience.
```

    Ces travaux ont été réalisés avant la sortie de PostgreSQL 17 qui ajoute d'importantes fonctionnalités pour le JSON avec les [`JSON_TABLE`](https://doc.postgresql.fr/17/functions-json.html#FUNCTIONS-SQLJSON-TABLE), elles ne seront pas évoquées ici.

> Puisque nous allons parler de json et de données semi-structurées, je me sens dans l'obligation de commencer cet article par un avertissement.

Reviewer: ```suggestion
Puisque nous allons parler de JSON et de données semi-structurées, je me sens dans l'obligation de commencer cet article par un avertissement.
```

**Le modèle relationnel, c'est bon, mangez-en, et les contraintes d'intégrités ont été inventées pour de bonnes raisons.**

> Cet article ne se veut surtout pas être une invitation à partir en mode yolo sur la gestion des données "c'est bon ya qu'a tout mettre en json" (comme un vulgaire dev qui mettrait tout dans MongoDB diraient les mauvaises langues).

Reviewer: ```suggestion
Cet article ne se veut surtout pas être une invitation à partir en mode YOLO sur la gestion des données "c'est bon ya qu'a tout mettre en JSON" (comme un vulgaire dev qui mettrait tout dans MongoDB diraient les mauvaises langues).
```

> ## Le JSON pour les débutant.es

Reviewer: ```suggestion
## Le JSON pour les débutant⸱es
```

Pour celles et ceux qui ne connaissent pas le `json`, il s'agit d'un format textuel de représentation des données venant du Java Script fonctionnant en partie sur un système de `clé : valeur` qu'on peut voir comme une sorte d'évolution du `xml`.

```json
{"clé_1": "valeur", "clé_2": "valeur", "clé_3": "valeur"}
```

> pas besoin de guillemets pour les nombres

Reviewer: ```suggestion
Pas besoin de guillemets pour les nombres :
```

```json
{"nb_champignons": 42, "nb_tomates": 31, "prenom": "roger"}
```

> Les valeurs peuvent prendre deux formes. Soit une valeur unique comme dans l'exemple ci-dessus, soit un `array`, une liste, qu'on place entre `[]`, les deux pouvant êtres combinés au sein d'un seul objet json.

Reviewer: ```suggestion
Les valeurs peuvent prendre deux formes :

- soit une valeur unique comme dans l'exemple ci-dessus,
- soit un `array`, une liste, qu'on place entre `[]`, les deux pouvant êtres combinés au sein d'un seul objet JSON.
```

Et les sous-objets ?

Reviewer: ok c'est dit juste après :innocent:

## Les types json de PostgreSQL

PostgreSQL est capable de stocker les données/objets au format json dans des champs auxquels on attribuera un type dédié. Il y en a deux car sinon ça ne serait pas rigolo.

- Le type `json` qui est là pour des raisons historiques et laisser fonctionner des bases qui auraient utilisé ce type dans le passé. Il stocke les informations sous forme textuelle, ce qui est peu optimisé pour un ordinateur. Il existe toutefois un intérêt à l'utiliser : il permet de retrouver l'information sur l'ordre des clés. S'il est important pour vous de savoir que `nom` est la clef 1 et `prenom` la clef 2, sans avoir à repasser par le nom de la clef, alors il vous faudra passer par le type `json`.

- le type `jsonb`. Le type moderne. Il stocke les informations sous forme binaire et énormément de fonctions sont disponibles en plus de celles du type `json`.

> ## Les index
> 
> Il est possible d'indexer un champ de type `json` / `jsonb` sur ses clés **de premier niveau** et cela se fait avec des index de type `GIN` :

Reviewer: ```suggestion
## Les index

![icône index](https://cdn.geotribu.fr/img/logos-icones/divers/index_pointeur.webp){: .img-thumbnail-left }

Il est possible d'indexer un champ de type `json` / `jsonb` sur ses clés **de premier niveau** et cela se fait avec des index de type `GIN` :
```

```sql
CREATE INDEX idx_tb_champjson ON tb USING gin (champ_json);
```

Pour indexer des valeurs plus "profondes", il faudra passer par des indexs fonctionnels, index sur des fonctions :

```sql
CREATE INDEX idx_tb_champjson ON tb USING gin (champ_json -> cle_ou_se_situent_les_valeur_a_indexer);
```

On expliquera ce `->` par la suite.

> Pour les aventuriers et aventurières, il existe une extention de PostgreSQL `btree_gin` permettant de faire des index multi-champs mixant des champs classiques et json. Elle est disponible nativement à l'installation de PostgreSQL et ne vous demandera pas de devenir développeur.se C/C++ pour l'installer (coucou les fdw parquet ! Ca va par chez vous ?).

Reviewer: ```suggestion
Pour les aventuriers et aventurières, il existe une extension de PostgreSQL `btree_gin` permettant de faire des index multi-champs mixant des champs classiques et `json`. Elle est disponible nativement à l'installation de PostgreSQL et ne vous demandera pas de devenir développeur⸱se C/C++ pour l'installer (coucou les FDW parquet ! Ça va par chez vous ?).
```

## Création des tables

> Je vais éviter de vous spammer du `DDL` ([Data Definition Language](https://fr.wikipedia.org/wiki/Langage_de_d%C3%A9finition_de_donn%C3%A9es), mais vous pourrez retrouver le schéma complet de la base [ici](https://github.com/thomas-szczurek/base_donnees_insee/tree/main/sql/creation_tables)

Reviewer: ```suggestion
Je vais éviter de vous spammer du `DDL` ([Data Definition Language](https://fr.wikipedia.org/wiki/Langage_de_d%C3%A9finition_de_donn%C3%A9es), mais vous pourrez retrouver le schéma complet de la base [ici](https://github.com/thomas-szczurek/base_donnees_insee/tree/main/sql/creation_tables).
```

En voici cependant un schéma succinct pour aider à la compréhension du reste de l'article :

> ![modele_de_donnees](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2024/postgresql_json/modele.png){: .img-center loading=lazy }

Reviewer: Pour ton karma envers les outils d'accessibilité :wink: 

```suggestion
![Diagramme du modèle de la base de données](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2024/postgresql_json/modele.png){: .img-center loading=lazy }
```

> Partant d'un schéma nommé `insee`, on va créer deux tables. La première contiendra la liste des *bases* disponibles, les différents volets du recensement. Une seconde permettant de stocker les données ; pour rester concentré sur le json, on va s'épargner 95% du modèle sous-jacent. On ne gérera donc pas ici les codes communes, etc. En bonus pour celles et ceux voulant tester ce que Postgres a dans le ventre, je proposerai une variante juste en dessous :

Reviewer: ```suggestion
Partant d'un schéma nommé `insee`, on va créer deux tables :+1: 
- la première contiendra la liste des *bases* disponibles, les différents volets du recensement ;
- lae seconde permettant de stocker les données.

Pour rester concentré sur le JSON, on va s'épargner 95% du modèle sous-jacent. On ne gérera donc pas ici les codes communes, etc. En bonus pour celles et ceux voulant tester ce que PostgreSQL a dans le ventre, je proposerai une variante juste en dessous :
```

   ELSE ((d.donnees ->> clef_json)::real)
  END AS valeur
 FROM tc
 JOIN
  insee.donnees_communes AS d ON (tc.code_com = d.code_commune AND tc.annee = d.annee)
 LEFT JOIN
  insee.presence_clefs_annees AS co ON tc.pk_id = co.pk_id
 ORDER BY tc.annee, co.fk_base, co.clef_json, tc.code_com
 )
SELECT * FROM final WHERE valeur IS NOT NULL;
```

![vm_presence_cles_annees](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2024/postgresql_json/vm_donnees_olap.png){: .img-center loading=lazy }

Attention, la création de cette vue matérialisée ou son refresh peut prendre un certain temps si vous avez stocké beaucoup de données (1 heure chez moi pour les 6 volets de 2015 à 2021).

Enfin, histoire de vivre avec son temps et non comme un viel ours des cavernes ou comme un data machin de collectivité territoriale (j'y bosse, j'ai le droit), on va convertir cette vue matérialisée en fichier [parquet](https://parquet.apache.org/).
Et, pour ça, on va utiliser gdal/ogr qui est décidément incroyable.

```sh
ogr2ogr -of parquet donnees_insee.parquet PG:"dbname='insee' shcema='insee' tables='donnees_communes_olap' user='nom_utilisateur' password='votre_mot_de_passe'"
```

> Et on peut ainsi mettre le fichier sur un espace cloud, comme [ici](https://donnees-insee.s3.fr-par.scw.cloud/donnees_insee_olap.parquet) ! Vous pouvez ensuite sortir votre plus beau générateur de publications Linkedin qui mettra plein d'emojis choupi et faire le cake sur les rezos (imaginez que 90% du contenu de Linkedin doit être fait avec ces trucs qui sont capable de vous générer des publications expliquant que le shape a des avantages sur le geopackage, le tout sur un ton très assuré).

Author: ```suggestion
Et on peut ainsi mettre le fichier sur un espace cloud, comme [ici](https://donnees-insee.s3.fr-par.scw.cloud/donnees_insee_olap.parquet) ! Vous pouvez ensuite sortir votre plus beau générateur de publications Linkedin qui mettra plein d'emojis choupi et faire le cake sur les rezos (imaginez que 90% du contenu de Linkedin doit être fait avec ces trucs qui sont capable de vous générer des publications expliquant que l'un  des rares avantages du shape sur le geopackage est d'être un format multi-fichiers, le tout sur un ton très assuré).
```

---
title: Travailler avec du JSON et PostgreSQL
subtitle: Jason et les éléphants
authors:
    - Thomas SZCZUREK-GAYANT
categories:
    - article
comments: true
date: 2025-01-21
description: "Stocker des données au format json dans PostgreSQL, les consulter... et tout ça avec les données du recensement de l'INSEE pour l'exemple. "
icon: simple/postgresql
image:
license: default
robots: index, follow
tags:
    - PostgreSQL
    - JSON
---

# Travailler avec du JSON dans PostgreSQL

:calendar: Date de publication initiale : {{ page.meta.date | date_localized }}

> Dans le cadre d'un projet personnel encadré de type "tu l'as fait pour les IRIS de Lille avec Sqlite, t'as vu que tu sais faire. Bon bah maintenant tu le fais pour toutes les communes de France dans PG", j'ai voulu stocker une bonne partie des données du recensement de l'Insee dans une base PostgreSQL avec des tables multimillésimes. Problème, même au sein d'un même jeu de données, les champs peuvent changer au cours des années et cela empêche de pouvoir dégager une structure de table fixe, ce qui est assez gênant vous en conviendrez. La solution ? Passer par des données semi-structurées, soit stocker ces données en JSON dans le champ d'une table. Cet article se veut un condensé de cette expérience.

Author: J'ai enlevé la partie entre guillemets pour alléger. C'était une ref à la personne qui m'a enseigné le sql "niveau ++" mais je comprend !

> !!! warning

Author: Pourquoi pas mais trop tard à deux jours près si je veux pas y passer mon dimanche :). J'espère que tu me pardonnera de choisir  les modifs que je veux encore faire.

    Ces travaux ont été réalisés avant la sortie de PostgreSQL 17 qui ajoute d'importantes fonctionnalités pour le JSON avec les [`JSON_TABLE`](https://doc.postgresql.fr/17/functions-json.html#FUNCTIONS-SQLJSON-TABLE), elles ne seront pas évoquées ici.

Puisque nous allons parler de json et de données semi-structurées, je me sens dans l'obligation de commencer cet article par un avertissement.

**Le modèle relationnel, c'est bon, mangez-en, et les contraintes d'intégrités ont été inventées pour de bonnes raisons.**

Cet article ne se veut surtout pas être une invitation à partir en mode yolo sur la gestion des données "c'est bon ya qu'a tout mettre en json" (comme un vulgaire dev qui mettrait tout dans Mongodb diraient les mauvaises langues).

## Le JSON pour les débutant.es

> Pour celles et ceux qui ne connaisent pas le `json`, il s'agit d'un format textuel de représentation des données fonctionnant en partie sur un système de `clé : valeur` qu'on peut voir comme une sorte d'évolution du `xml`.

Author: ```suggestion
Pour celles et ceux qui ne connaissent pas le `json`, il s'agit d'un format textuel de représentation des données venant du Java Script fonctionnant en partie sur un système de `clé : valeur` qu'on peut voir comme une sorte d'évolution du `xml`.
```

- le type `jsonb`. Le type moderne. Il stocke les informations sous forme binaire et énormément de fonctions sont disponibles en plus de celles du type `json`.

## Les index

Il est possible d'indexer un champ de type `json` / `jsonb` sur ses clés **de premier niveau** et cela se fait avec des index de type `GIN` :

```sql
CREATE INDEX idx_tb_champjson ON tb USING gin (champ_json);
```

Pour indexer des valeurs plus "profondes", il faudra passer par des indexs fonctionnels, index sur des fonctions :

```sql
CREATE INDEX idx_tb_champjson ON tb USING gin (champ_json -> cle_ou_se_situent_les_valeur_a_indexer);
```

On expliquera ce `->` par la suite.

Pour les aventuriers et aventurières, il existe une extention de PostgreSQL `btree_gin` permettant de faire des index multi-champs mixant des champs classiques et json. Elle est disponible nativement à l'installation de PostgreSQL et ne vous demandera pas de devenir développeur.se C/C++ pour l'installer (coucou les fdw parquet ! Ca va par chez vous ?).

## Création des tables

> Je vais éviter de vous spammer du DDL, mais vous pourrez retrouver le schéma complet de la base [ici](https://github.com/thomas-szczurek/base_donnees_insee/tree/main/sql/creation_tables)

Author: ```suggestion
Je vais éviter de vous spammer du `DDL` ([Data Definition Language](https://fr.wikipedia.org/wiki/Langage_de_d%C3%A9finition_de_donn%C3%A9es), mais vous pourrez retrouver le schéma complet de la base [ici](https://github.com/thomas-szczurek/base_donnees_insee/tree/main/sql/creation_tables)
```

> En voici cependant un schéma succinct pour pour aider à la compréhension du reste de l'article :

Author: ```suggestion
En voici cependant un schéma succinct pour aider à la compréhension du reste de l'article :
```

```sql
BEGIN;
-- table listant les differentes bases Insee disponibles

CREATE TABLE insee.bases (
pk_id int2 PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY,
nom text NOT NULL
);

-- table stockant les données par communes

CREATE TABLE insee.donnees_communes (
pk_id int4 PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY,
code_commune text NOT NULL CHECK length(code_commune) = 5, -- n'oublions pas les Corses qui ont des codes commune en 2A et 2B d'où le type text
fk_base int2 NOT NULL,
annee int2 NOT NULL,
donnees jsonb NOT NULL,
UNIQUE (fk_base, annee, code_commune)
CONSTRAINT fk_donnees_bases FOREIGN KEY fk_base ON insee.bases(pk_id) ON UPDATE CASCADE ON DELETE CASCADE
);

-- Création des index
-- Un index multichamps est déjà créé automatiquement sur (fk_base, annee, code_commune) grace à la clause UNIQUE de la déclaration de table
CREATE INDEX idx_donnees_com_donnees ON insee.donnees_communes USING gin (donnees);
END;
```

> Les plus tatillons des DBA d'entre vous auront remarqué ce `CHECK`. "Mais pourquoi qu'il utilise pas varchar(5) ? Il fait vraiment n'importe quoi !" Tout simplement, car cette forme permet d'utiliser un type de texte aux nombres de caractère **réellement** arbitraire (le type `text`) contrairement à varchar(255), tout en pouvant en contrôler le nombre minimum et maximum avec le prédicat `CHECK` (contrairement à varchar qui ne contrôle que le maximum) comme expliqué sur le [wiki Postgres](https://wiki.postgresql.org/wiki/Don%27t_Do_This#Don.27t_use_varchar.28n.29_by_default).

Author: C'est à ça que sert la table code_admin (et gérer les codes en doublons entre dpt et regions) un peu bancale du modèle ;)

Et on insère quelques lignes dans notre table insee.bases

```sql
INSERT INTO insee.bases (nom) VALUES
('rp_population'),
('rp_menages'),
('rp_logement'),
('rp_diplomes'),
('rp_activite'),
('rp_emploi')
```

> ### (Bonus) Partition de la table données

Author: Ca me donnerai l'impression de faire un article qui pourrait se résumer par un RTFM mais pour quoi pas en effet

('rp_population'),
('rp_menages'),
('rp_logement'),
('rp_diplomes'),
('rp_activite'),
('rp_emploi')
```

- Cela peut être intéressant pour limiter la taille des scans séquentiels qui se feront sur des tables plus petites que la table *parent*, où pour se faciliter la gestion de tables volumineuses.
- On peut attacher / détacher une partition (qui devient alors une table classique) avec les "mots" `ATTACH PARTITION` / `DETACH PARTITION` y compris sur une base en production grâce à l'option `CONCURENTLY`
- Une clé primaire de table partitionnée **doit** être composite, c'est-à-dire que son unicité sera vérifiée par la composition de plusieurs champs et contenir obligatoirement le champ de partitionnement.

> 

Author: ```suggestion
```

> Vous pouvez aussi lire le contenu de la [documentation à ce sujet](https://doc.postgresql.fr/17/ddl-partitioning.html)

Author: ```suggestion
```

> 

Author: ```suggestion
```

> À quoi ressemblerait la création de insee.donnees_communes si on la partitionnait selon les différentes sources de données ?

Author: ```suggestion
```

```sql

> BEGIN;

Author: ```suggestion
```

> CREATE TABLE insee.donnees_communes (

Author: ```suggestion
```

>  pk_id int4 GENERATED BY DEFAULT AS IDENTITY,

Author: ```suggestion
```

>  code_commune text NOT NULL CHECK length(code_commune) = 5,

Author: ```suggestion
```

 annee int2 NOT NULL,
 fk_base int2 NOT NULL,
 donnees jsonb NOT NULL,
 UNIQUE (annee, code_commune, fk_base),
 CONSTRAINT pk_donnees_communes PRIMARY KEY (pk_id, fk_base),
 CONSTRAINT fk_donnees_bases FOREIGN KEY (fk_base) REFERENCES insee.bases (pk_id) ON UPDATE CASCADE ON DELETE CASCADE
) PARTITION BY LIST (fk_base);
CREATE TABLE insee.donnees_communes_fk_1 PARTITION OF insee.donnees_communes FOR VALUES IN (1);
CREATE TABLE insee.donnees_communes_fk_2 PARTITION OF insee.donnees_communes FOR VALUES IN (2);
CREATE TABLE insee.donnees_communes_fk_3 PARTITION OF insee.donnees_communes FOR VALUES IN (3);
CREATE TABLE insee.donnees_communes_fk_4 PARTITION OF insee.donnees_communes FOR VALUES IN (4);
CREATE TABLE insee.donnees_communes_fk_5 PARTITION OF insee.donnees_communes FOR VALUES IN (5);
CREATE TABLE insee.donnees_communes_fk_6 PARTITION OF insee.donnees_communes FOR VALUES IN (6);
END;
```

> On remarque quelques changements : la déclaration de la clé primaire composite dans une `CONSTRAINT`, le `PARTITION BY LIST` qu'on fait suivre par le champ de partitionnement, et la création des différentes partitions suivant les valeurs de ce champ.

Author: ```suggestion
```

### Exemple simple

On va créer une chaine de texte qui contiendra le contenu de notre json dont les valeurs seront séparées par des virgules `clé1,valeur1, clé2, valeur2`. Cette chaine sera passée dans une fonction `string_to_array()` la transformant en `array` avec comme séparateur des `,` pour séparer les éléments de la chaine de texte vers des éléments de liste, caractère passé en second paramètre de la fonction. Cet `array` sera ensuite envoyé dans la fonction `jsonb_object()`.

```sql
INSERT INTO insee.donnees_communes (code_commune, annee, fk_base, donnees) VALUES
(
  '99999',
  2024,
  1,
  jsonb_object(
    string_to_array('tomates,42,melons,12',',')
    )
)
```

Cette requête encodera cet objet json dans le champs `données` :

```json
{
  "tomate": 42,
  "melons": 12
}
```

> Maintenant comment récupérer notre nombre de melons pour le code commune 99999 en 2024 ? Celà se fait grace à des opérateurs spéciaux :

Author: pour suivre le schéma de la table, mais du coup j'ai complété le WHERE pour faire correspondre la question exacte au sql

- `champ_jsonb -> 'clé'` récupère la valeur d'une clé en conservant son type json
- `champ_jsonb ->> 'clé'` fait de même en transformant la valeur en type sql "classique"

```sql
SELECT
  (donnees ->> 'melons')::int4 AS nb_melons
FROM insee.donnees_communes
WHERE

>   donnees ? 'melons'

Author: ```suggestion
  donnees ? 'melons'
  AND code_commune = '99999'
  AND annee = 2024
```

### Avec du json imbriqué

Bon, il est bien gentil, mais là son json reste du json où tout est au premier niveau. Bah oui, pour mon besoin, cela m'a suffi. Mais, je ne vais pas fuir mes responsabilités et on va voir comment cela se passe avec du json plus complexe. Je repartirai de l'exemple de la partie précédente pour complexifier après cet aparté.

Pour injecter du json complexe dans un champ, deux solutions s'offrent à nous : imbriquer les fonctions dédiées, ou caster une chaine de texte. Imaginons une table "test", dans un schema "test" et dont un champ `jsonb` se nomme "donnees".

```sql
INSERT INTO test.test (donnees) VALUES
(jsonb_object(
 'cle1', 'valeur1',
 'cle2', jsonb_array(
  'foo', 'bar', 'baz')
  )
 )
```

Cette insertion pourrait tout aussi bien s'écrire avec un cast d'une chaine de texte vers du jsonb. Attention, la syntaxe json doit être ici respectée :

```sql
INSERT INTO test.test (donnees) VALUES
('{"cle1": "valeur1", "cle2": ["foo", "bar", "baz"]}'::jsonb)
```

> 

Author: ```suggestion

#### Recupération

```

Pour récupérer une valeur, on utilise la fonction `jsonb_path_query()` qui possède deux paramètres : `le nom du champ` contenant les données json, et le `json_path` vers la valeur à atteindre. Imaginons que nous voulions récupérer la deuxième valeur de la liste contenue dans "cle2" :

```sql
SELECT
 jsonb_path_query(donnee, '$.cle2[1]')
FROM test.test
```

Le `$` désigne le début du chemin json retourné. Nous faisons suivre ce premier symbole par un point pour passer à l'objet suivant. Puis, par le nom de clé suivant et ainsi de suite jusqu'à la clé recherchée, à laquelle nous collons un `[1]` pour la 2ᵉ valeur de la liste (les valeurs commencent à 0).
Pour plus d'informations sur les `json_path`, vous pouvez consulter la [documentation](https://www.postgresql.org/docs/current/functions-json.html#FUNCTIONS-SQLJSON-PATH)

## Attaquons nous au recensement

Bien, maintenant que nous avons essayé d'expliquer tant bien que mal les concepts avec des exemples simples, car il faut bien commencer quelque part, essayons avec quelque chose d'un peu plus volumineux.

> Récupérons le dernier millésime du volet population du recensement communal [au format csv sur le site de l'INSEE](https://www.insee.fr/fr/information/8183122) (vous voulez les bases des principaux indicateurs). Pour l'exemple, on utilisera le fichier "Evolution et structure de la population". Il faut tout d'abord nettoyer les noms de champ. En effet, l'INSEE indique le millésime systématiquement dans ses noms de champ, ce qui fait que ces derniers changent tous les ans pour un même indicateur.

Reviewer 3: ```suggestion
Récupérons le dernier millésime du volet population du recensement communal [au format csv sur le site de l'Insee](https://www.insee.fr/fr/information/8183122) (vous voulez les bases des principaux indicateurs). Pour l'exemple, on utilisera le fichier "Evolution et structure de la population". Il faut tout d'abord nettoyer les noms de champ. En effet, l'Insee indique le millésime systématiquement dans ses noms de champ, ce qui fait que ces derniers changent tous les ans pour un même indicateur.
```

Par cohérence avec plus haut...et même si c'est des initiales, je vois que l'Insee écrit Insee sur son site.

> Les noms de champ commencent tous par P ou C, ceci indique *exploitation principale* (réponses brutes aux questions du recensement) ou *exploitation complémentaire* (croisement de réponses pour établir un indicateur). Les champs provenant de l'exploitation principale et ceux issus de la complémentaire ne doivent pas être croisés entre eux. Cette information est évidemment à conserver, mais par choix personnel, je préfère la mettre à la fin plutôt qu'au début du nom.

Reviewer 3: Je trouve ce qui suit, sur la gestion des noms de colonnes, plutôt éloigné de l'objet principal de l'article.
Je dirais qu'il faut diminuer, peut-être juste évoquer ce pré-traitement sans entrer dans les détails.

Inversement, si tu préfères laisser cette section, je la trouve pas assez explicite, quelques petites captures avant/après pourraient aider à comprendre chacune des étapes.

Author: L'information sur les rincipales/complémentaire mon semble un bon point de culture gé pour des gens qui passent leur temps devant ce type de données, mais j'ai enlevé tout le passage Excel, en ajoutant juste sur ligne sur le produit a obtenir.

> Les noms de champ commencent tous par P ou C, ceci indique *exploitation principale* (réponses brutes aux questions du recensement) ou *exploitation complémentaire* (croisement de réponses pour établir un indicateur). Les champs provenant de l'exploitation principale et ceux issus de la complémentaire ne doivent pas être croisés entre eux. Cette information est évidemment à conserver, mais par choix personnel, je préfère la mettre à la fin plutôt qu'au début du nom.

Author: ```suggestion
Les noms de champ commencent tous par P ou C, ceci indique *exploitation principale* (réponses brutes aux questions du recensement) ou *exploitation complémentaire* (croisement de réponses pour établir un indicateur). Les champs provenant de l'exploitation principale et ceux issus de la complémentaire ne doivent pas être croisés entre eux. Cette information est évidemment à conserver, mais par choix personnel, je préfère la mettre à la fin plutôt qu'au début du nom. On passera ainsi de champs normés comme ceci  : `P18_POP` vers une normalisation de ce type `POP_P`.
```

Vous trouverez [ici](https://github.com/thomas-szczurek/base_donnees_insee/blob/main/sql/import/correction_champs_insee.xlsx) un tableur dont le rôle est de s'occuper de tout ça.

> - `STXT` est la fonction qui permet de découper une chaine de texte avec Excel, avec comme arguments `cellule comprenant la chaine initiale` , `position du début de la découpe` , `position de fin de la découpe`

Author: ```suggestion
```

> - À partir du fichier récupéré, il vous suffit de copier-coller les noms de champs INSEE sur la première ligne et de récupérer la seconde, qui contiendra les noms de champs formatés.

Author: ```suggestion
```

> - On  remplace les noms de champ du fichier INSEE original.

Author: ```suggestion
```

> - Et on change le nom de cette table temporaire par "rp_population_import.csv"

Author: ```suggestion
```

> Avant d'insérer les données dans notre table, nous allons passer par une table temporaire afin de rendre les données accessibles dans Postgres. Utiliser `COPY` de Postgresql serait fastidieux car il faudrait indiquer la centaine de champs que contient le volet population du recensement dans la commande. Et, je n'ai pas honte de dire que j'ai un baobab dans la main à cette idée. Nous sortons donc ce merveilleux logiciel qu'est QGIS. On active les panneaux Explorateur et Explorateur2. On crée une connexion vers la base avec les droits de création, et d'un mouvement gracile du poignet, vous glissez le fichier depuis le panneau Explorateur vers la base Postgres dans l'Explorateur2. Laisser la magie opérer.

Reviewer 3: ```suggestion
Avant d'insérer les données dans notre table, nous allons passer par une table temporaire afin de rendre les données accessibles dans Postgres. Utiliser `COPY` de Postgresql serait fastidieux car il faudrait indiquer la centaine de champs que contient le volet population du recensement dans la commande. Et, je n'ai pas honte de dire que j'ai un baobab dans la main à cette idée. Nous sortons donc ce merveilleux logiciel qu'est QGIS. On active les panneaux Explorateur et Explorateur2. On crée une connexion vers la base avec les droits de création, et d'un mouvement gracile du poignet, vous glissez le fichier depuis le panneau Explorateur vers la base Postgres dans l'Explorateur2. Laissez la magie opérer.
```

> Maintenant, préparez-vous pour peut-être un des INSERT les plus bizarres de votre vie (en tout cas, ça l'a été pour moi !). Arf. Je me rends compte que si je veux bien faire, il faut aussi que j'explique les CTE (c'est très étrange, car encore une fois, si vous êtes encore ici, vous savez très probablement déjà ce qu'est une CTE).

Author: ```suggestion
Maintenant, préparez-vous pour peut-être un des INSERT les plus bizarres de votre vie (en tout cas, ça l'a été pour moi !). Arf. Je me rends compte que si je voulais bien faire, il faudrait aussi que j'explique les [CTE](https://www.postgresql.org/docs/current/queries-with.html). Mais pour ne pas trop alourdir, je vous laisser cliquer sur le lien.
```

> CTE veut dire Common Table Expression. C'est une fonctionnalité qui permet grâce à la clause `WITH` d'isoler une sous-requête de la requête principale pour rendre tout un peu plus clair. Elle permet également de la nommer pour pouvoir la réutiliser à plusieurs endroits sans devoir la réécrire. On peut aussi s'en servir pour faire des requêtes récursives avec `WITH RECURSIVE` et si le sujet vous intéresse je vous encourage à aller lire la [magnifique documentation](https://www.postgresql.org/docs/current/queries-with.html#QUERIES-WITH-RECURSIVE) de Postgres a ce sujet.

Author: ```suggestion
```

On va utiliser la CTE pour concaténer le nom que l'on veut donner à nos clés avec les valeurs contenues dans notre table temporaire dans une chaine séparée par des `,`. On l'enverra dans une fonction `string_to_array()` puis dans une fonction `jsonb_object()`. On en profitera pour au passage, nettoyer toute tabulation ou retour chariot qui pourrait subsister avec une expression régulière grâce à la fonction `regex_replace()`. (ces caractères se notent `\t`, `\n` et `\r`). Cette dernière fonction prend 3 arguments : la chaine de caractères source, le `pattern` recherché, le texte de remplacement. On y ajoute le *drapeau* optionnel `g` afin de remplacer toutes les occurrences trouvées.

Notez que si votre table temporaire possède un nom différent de "rp_population_import" il vous faudra modifier la clause FROM de la CTE.

```sql
-- cte concatenant les données avec les clés et nettoyant les caractères spéciaux.
WITH d AS (
  SELECT

>  "CODGEO",

Author: ```suggestion
 CODGEO,
```

>  regexp_replace('pop_p,' || "POP_P" || ',

Reviewer 3: Ce serait plus simple avec un concat ou concat_ws qu'avec les || non ?

Author: Je vois l'idée, mais c'est comme ça que je l'ai écrite à l'époque (j'ai appris dessus) et là pareil, trop tard pour changer

   pop_25_54_ans_cadres_prof_intel_sup_c,' || "POP2554_CS3_C" || ',
   pop_25_54_ans_professions_intermediaires_c,' || "POP2554_CS4_C" || ',
   pop_25_54_ans_employes_c,' || "POP2554_CS5_C" || ',
   pop_25_54_ans_ouvriers_c,' || "POP2554_CS6_C" || ',
   pop_25_54_ans_retraites_c,' || "POP2554_CS7_C" || ',
   pop_25_54_ans_autres_c,' || "POP2554_CS8_C" || ',
   population_55_ans_et_plus_c,' || "POP55P_C" || ',
   pop_55_ans_et_plus_ans_agriculteurs_c,' || "POP55P_CS1_C" || ',
   pop_55_ans_et_plus_ans_artisants_commercants_chefs_entreprise_c,' || "POP55P_CS2_C" || ',
   pop_55_ans_et_plus_ans_cadres_prof_intel_sup_c,' || "POP55P_CS3_C" || ',
   pop_55_ans_et_plus_ans_professions_intermediaires_c,' || "POP55P_CS4_C" || ',
   pop_55_ans_et_plus_ans_employes_c,' || "POP55P_CS5_C" || ',
   pop_55_ans_et_plus_ans_ouvriers_c,' || "POP55P_CS6_C" || ',
   pop_55_ans_et_plus_ans_retraites_c,' || "POP55P_CS7_C" || ',
   pop_55_ans_et_plus_ans_autres_c,' || "POP55P_CS8_C",
  E'[\t\n\r]','','g') AS data
  FROM insee.rp_population_import
 )

-- Conversion des chaines de texte en json et insertion dans la table
INSERT INTO insee.donnees_communes("code_commune","annee","fk_base","donnees")
SELECT

>     "CODGEO",

Author: ```suggestion
    CODGEO,
```

    2021,
    1,
    jsonb_object(string_to_array(data::text,','))
FROM d
ORDER BY "CODGEO";
```

> Ouf.

Reviewer 3: J'ai vraiment du mal à suivre ce qui précède.

En plus, le markdown en preview colore de la même façon les 'chaines' et les "CHAINES" alors que les simples et doubles guillemets ne représente pas la même chose en PG (contrairement en Python).
Ainsi, j'ai mis un temps à voir que select "CODGEO" n'est pas une sélection de constante.

Author: J'ai au moins enlevé ceux de CODGEO pour la comprehension, je vois si je peux changer les autres en masse

![donnees_communes](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2024/postgresql_json/donnees_communes.png){: .img-center loading=lazy }

> On veillera à bien grouper les inserts dans cet ordre annee/base/code_commune afin de faciliter la lecture des données par PostgreSQL (si votre table est partitionnée, celà sera facilité).

Reviewer 3: ```suggestion
On veillera à bien grouper les inserts dans cet ordre annee/base/code_commune afin de faciliter la lecture des données par PostgreSQL (si votre table est partitionnée, cela sera facilité).
```

## Et maintenant ? Qu'est-ce qu'on fait de ça ?

Jusqu'à maintenant, nous n'avons travaillé qu'avec le volet population pour son dernier millésime. Imaginons maintenant que nous répétions l'exercice pour les 6 volets et sur plusieurs millésimes, et ce, en sachant qu'au cours du temps, certains champs peuvent apparaitre ou disparaitre ; changement dans les niveaux de diplômes observés par exemple. Il serait intéressant de récupérer une table indiquant la première et la dernière année de présence de chaque clé. Mettons que lors de ces travaux, nous en avons profité pour mettre à jour une table "correspondance_clefs_champs" listant chaque clé présentes et son nom INSEE d'origine (tout du moins, celui que nous avions normalisé).

```sql
CREATE MATERIALIZED VIEW insee.presence_clefs_annees AS
SELECT
 c.pk_id AS pk_id,
 c.clef_json AS clef_json,
 c.fk_base AS fk_base,

>  p.premiere AS premiere_annee_presence,

Author: ```suggestion
 a.premiere AS premiere_annee_presence,
```

>  d.derniere AS derniere_annee_presence

Author: ```suggestion
 a.derniere AS derniere_annee_presence
```

FROM insee.correspondance_clefs_champs AS c,

>  LATERAL (SELECT min(annee) AS premiere FROM insee.donnees_communes WHERE fk_base = c.fk_base AND donnees ? c.clef_json) AS p,

Reviewer 3: J'ai un doute, on peut pas faire le min et max dans la même sous-requêtes ?

Author: Si, et on gagne même 211 ms d’exécution, hourra ! Modifié, merci !

>  LATERAL (SELECT min(annee) AS premiere FROM insee.donnees_communes WHERE fk_base = c.fk_base AND donnees ? c.clef_json) AS p,

Author: ```suggestion
 LATERAL (SELECT min(annee) AS premiere FROM insee.donnees_communes WHERE fk_base = c.fk_base AND donnees ? c.clef_json) AS a
```

>  LATERAL (SELECT max(annee) AS derniere FROM insee.donnees_communes WHERE fk_base = c.fk_base AND donnees ? c.clef_json) AS d

Author: ```suggestion
```

  FROM insee.presence_clefs_annees
 ),
-- cross join des clefs + année unnestées et des codes communaux
 tc AS (
 SELECT
  cc.code_admin AS code_com,
  cl.annee AS annee,
  cl.pk_id AS pk_id
 FROM codes_com AS cc
 CROSS JOIN clefs AS cl
 ),
-- Selection finale avec récupération des données
 final AS (
 SELECT
  tc.code_com,
  tc.annee,
  co.fk_base,
  co.clef_json,
  CASE
   WHEN (d.donnees ->> clef_json) IN ('','null','s','nd') THEN NULL
   ELSE ((d.donnees ->> clef_json)::real)
  END AS valeur
 FROM tc
 JOIN
  insee.donnees_communes AS d ON (tc.code_com = d.code_commune AND tc.annee = d.annee)
 LEFT JOIN
  insee.presence_clefs_annees AS co ON tc.pk_id = co.pk_id
 ORDER BY tc.annee, co.fk_base, co.clef_json, tc.code_com
 )
SELECT * FROM final WHERE valeur IS NOT NULL;
```

![vm_presence_cles_annees](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2024/postgresql_json/vm_donnees_olap.png){: .img-center loading=lazy }

Attention, la création de cette vue matérialisée ou son refresh peut prendre un certain temps si vous avez stocké beaucoup de données (1 heure chez moi pour les 6 volets de 2015 à 2021).

> Enfin, histoire de vivre avec son temps et non comme un viel ours des cavernes, on va convertir cette vue matérialisée en fichier [parquet](https://parquet.apache.org/).

Reviewer 3: ```suggestion
Enfin, histoire de vivre avec son temps et non comme un vieil ours des cavernes, on va convertir cette vue matérialisée en fichier [parquet](https://parquet.apache.org/).
```

Et, pour ça, on va utiliser gdal/ogr qui est décidément incroyable.

```sh
ogr2ogr -of parquet donnees_insee.parquet PG:"dbname='insee' shcema='insee' tables='donnees_communes_olap' user='nom_utilisateur' password='votre_mot_de_passe'"
```

> Et on peut ainsi mettre le fichier sur un espace cloud, comme [ici](https://donnees-insee.s3.fr-par.scw.cloud/donnees_insee_olap.parquet) ! Vous pouvez ensuite sortir votre plus beau générateur de publications Linkedin qui mettra plein d'emojis choupi et faire le cake sur les rezos (imaginez que 90% du contenu de Linkedin doit être fait avec ces trucs qui sont capable de vous générer des publications expliquant que l'un  des rares avantages du shape sur le geopackage est d'être un format multi-fichiers, le tout sur un ton très assuré).

Author: ```suggestion
Et on peut ainsi mettre le fichier sur un espace cloud, comme [ici](https://donnees-insee.s3.fr-par.scw.cloud/donnees_insee_olap.parquet) ! Vous pouvez ensuite sortir votre plus beau générateur de publications Linkedin qui mettra plein d'emojis choupi et faire le cake sur les rezos (imaginez que 90% du contenu de Linkedin doit être fait avec ces trucs qui sont capable de vous générer des publications expliquant que l'un des rares avantages du shape sur le geopackage est d'être un format multi-fichiers, le tout sur un ton très assuré).
```

```

Pour indexer des valeurs plus "profondes", il faudra passer par des indexs fonctionnels, index sur des fonctions :

```sql
CREATE INDEX idx_tb_champjson ON tb USING gin (champ_json -> cle_ou_se_situent_les_valeur_a_indexer);
```

On expliquera ce `->` par la suite.

Pour les aventuriers et aventurières, il existe une extention de PostgreSQL `btree_gin` permettant de faire des index multi-champs mixant des champs classiques et json. Elle est disponible nativement à l'installation de PostgreSQL et ne vous demandera pas de devenir développeur.se C/C++ pour l'installer (coucou les fdw parquet ! Ca va par chez vous ?).

## Création des tables

Je vais éviter de vous spammer du `DDL` ([Data Definition Language](https://fr.wikipedia.org/wiki/Langage_de_d%C3%A9finition_de_donn%C3%A9es), mais vous pourrez retrouver le schéma complet de la base [ici](https://github.com/thomas-szczurek/base_donnees_insee/tree/main/sql/creation_tables)

En voici cependant un schéma succinct pour aider à la compréhension du reste de l'article :

![modele_de_donnees](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2024/postgresql_json/modele.png){: .img-center loading=lazy }

> 

Author: ```suggestion

*Ne vous préoccupez pas des tables empilées tout à droite, ce sont des partitions de la table `donnees_communes`. Le sujet ne sera pas évoqué ici et elles ne sont nécessaires pour l'exemple.*

```

UNIQUE (fk_base, annee, code_commune)
CONSTRAINT fk_donnees_bases FOREIGN KEY fk_base ON insee.bases(pk_id) ON UPDATE CASCADE ON DELETE CASCADE
);

-- Création des index
-- Un index multichamps est déjà créé automatiquement sur (fk_base, annee, code_commune) grace à la clause UNIQUE de la déclaration de table
CREATE INDEX idx_donnees_com_donnees ON insee.donnees_communes USING gin (donnees);
END;
```

Les plus tatillons des DBA d'entre vous auront remarqué ce `CHECK`. "Mais pourquoi qu'il utilise pas varchar(5) ? Il fait vraiment n'importe quoi !" Tout simplement, car cette forme permet d'utiliser un type de texte aux nombres de caractère **réellement** arbitraire (le type `text`) contrairement à varchar(255), tout en pouvant en contrôler le nombre minimum et maximum avec le prédicat `CHECK` (contrairement à varchar qui ne contrôle que le maximum) comme expliqué sur le [wiki Postgres](https://wiki.postgresql.org/wiki/Don%27t_Do_This#Don.27t_use_varchar.28n.29_by_default).

Et on insère quelques lignes dans notre table insee.bases :

```sql
INSERT INTO insee.bases (nom) VALUES
('rp_population'),
('rp_menages'),
('rp_logement'),
('rp_diplomes'),
('rp_activite'),
('rp_emploi')
```

> ### (Bonus) Partition de la table données

Author: ```suggestion
```

> Imaginons que nous travaillons sur l'ensemble des données du recensement (soit 6 fichiers sources), de 2015 à 2021 pour environ 35 000 communes. On va arriver sur du 1,5 million d'enregistrements. Entendons-nous bien, à l'échelle de Postgres ça ne reste pas grand-chose. Mais, si comme moi, vous voulez voir ce qu'on peut tirer des entrailles de Postres, ça permet de pouvoir commencer à justifier d'utiliser certaines fonctionnalités avancées. Une table partitionnée, c'est quoi ?

Author: ```suggestion
```

> 

Author: ```suggestion
```

> - C'est une table découpée en plusieurs morceaux où une table *parent* contrôlera les tables *enfants*.

Author: ```suggestion
```

> - Une table partitionnée peut se découper selon la valeur d'un champ ou une période temporelle ("fais -moi une partition tous les mois")

Author: ```suggestion
```

> - On peut requêter la table parent ou les tables enfants.

Author: ```suggestion
```

> - Une table doit être partitionnée à sa création, une table déjà existante ne peut pas être convertie. De nouvelles partitions peuvent toutefois être créées à volonté.

Author: ```suggestion
```

## Et maintenant ? Qu'est-ce qu'on fait de ça ?

Jusqu'à maintenant, nous n'avons travaillé qu'avec le volet population pour son dernier millésime. Imaginons maintenant que nous répétions l'exercice pour les 6 volets et sur plusieurs millésimes, et ce, en sachant qu'au cours du temps, certains champs peuvent apparaitre ou disparaitre ; changement dans les niveaux de diplômes observés par exemple. Il serait intéressant de récupérer une table indiquant la première et la dernière année de présence de chaque clé. Mettons que lors de ces travaux, nous en avons profité pour mettre à jour une table "correspondance_clefs_champs" listant chaque clé présentes et son nom INSEE d'origine (tout du moins, celui que nous avions normalisé).

```sql
CREATE MATERIALIZED VIEW insee.presence_clefs_annees AS
SELECT
 c.pk_id AS pk_id,
 c.clef_json AS clef_json,
 c.fk_base AS fk_base,
 a.premiere AS premiere_annee_presence,
 a.derniere AS derniere_annee_presence
FROM insee.correspondance_clefs_champs AS c,

>  LATERAL (SELECT min(annee) AS premiere FROM insee.donnees_communes WHERE fk_base = c.fk_base AND donnees ? c.clef_json) AS a

Author: ```suggestion
 LATERAL (SELECT
     min(annee) AS premiere,
     max(annee) AS derniere 
     FROM insee.donnees_communes WHERE fk_base = c.fk_base AND donnees ? c.clef_json) AS a
```

ORDER BY fk_base, clef_json;
```

![vm_presence_cles_annees](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2024/postgresql_json/vm_presence_cles_annees.png){: .img-center loading=lazy }

> La seule difficulté ici est la présence de [LATERAL](https://docs.postgresql.fr/17/queries-table-expressions.html#QUERIES-FROM). Ce « mot » permet d'utiliser un champ de la requête principale dans une sous-requête placée dans une clause FROM. Les éléments de la sous-requête seront ensuite évalués atomiquement avant d'être joints à la table d'origine. Oui, ce n'est pas très facile à expliquer. Ici, le WHERE de la sous-requête va interroger le champ donné de la table "donnees_communes" pour voir s'il y voit la clef_json en train d'être évaluée dans la table "correspondance_clefs_champs" possédant l'alias "c". Si oui, alors on prend la valeur minimum/maximum du champ année et on le joint à cette ligne et uniquement à cette ligne. Puis évaluation de clef_json suivante ... (*évaluée atomiquement*)

Author: ```suggestion
La seule difficulté ici est la présence de [LATERAL](https://docs.postgresql.fr/17/queries-table-expressions.html#QUERIES-FROM). Ce « mot » permet d'utiliser un champ de la requête principale dans une sous-requête placée dans une clause FROM. Les éléments de la sous-requête seront ensuite évalués atomiquement avant d'être joints à la table d'origine. Oui, ce n'est pas très facile à expliquer. Ici, le WHERE de la sous-requête va interroger le champ `donnees` de la table "donnees_communes" pour voir s'il y voit la clef_json en train d'être évaluée dans la table "correspondance_clefs_champs" possédant l'alias "c". Si oui, alors on prend la valeur minimum/maximum du champ année et on le joint à cette ligne et uniquement à cette ligne. Puis évaluation de clef_json suivante ... (*évaluée atomiquement*)
```

}
```

Maintenant comment récupérer notre nombre de melons pour le code commune 99999 en 2024 ? Celà se fait grace à des opérateurs spéciaux :

- `champ_jsonb -> 'clé'` récupère la valeur d'une clé en conservant son type json
- `champ_jsonb ->> 'clé'` fait de même en transformant la valeur en type sql "classique"

```sql
SELECT
  (donnees ->> 'melons')::int4 AS nb_melons
FROM insee.donnees_communes
WHERE
  donnees ? 'melons'
  AND code_commune = '99999'
  AND annee = 2024
```

Vous pouvez voir que j'utilise l'opérateur `?`; uniquement valable pour les champs `jsonb` et non ceux en simple `json`. En effet, lorsque l'on requête un champ json/jsonb, un retour vous est fait pour l'ensemble des enregistrements de la table, même ceux ne contenant pas la clé. Comprendre que si votre table contient 100 000 enregistrements, mais que seuls 100 contiennent la clé "melons", ne pas spécifier cette clause WHERE vous renverrait 100 000 lignes, dont 99 900 de `NULL`.`?` est un opérateur json permettant de poser la question "la clé est-elle présente au premier niveau du champ json pour cet enregistrement ?", et on ne récupérerait que nos 100 enregistrements contenant la clé "melons".

> Je précise que même si vous êtes encore ici, je suppose que vous savez déjà cela. Cependant, la forme `(quelque_chose)::int4` est un raccourci de PostgreSQL pour faire un `cast` soit convertir une valeur dans un autre type. Avec `->>` la valeur nous est renvoyée sous forme de texte et nous la convertissons en entier.

Author: ```suggestion
Je précise que même si vous êtes encore ici, je suppose que vous savez déjà cela. Cependant, la forme `(quelque_chose)::(type)` est un raccourci de PostgreSQL pour faire un `cast` soit convertir une valeur dans un autre type. Avec `->>` la valeur nous est renvoyée sous forme de texte et nous la convertissons en entier.
```
